src/osd/ReplicatedBackend.cc

```diff
@@ -1665,24 +1693,44 @@ void ReplicatedBackend::submit_push_data(
   map<string, bufferlist> &omap_entries,
   ObjectStore::Transaction *t)
 {
-  hobject_t target_oid;
-  if (first && complete) {
-    target_oid = recovery_info.soid;
-  } else {
-    target_oid = get_parent()->get_temp_recovery_object(recovery_info.version,
+  hobject_t target_oid = get_parent()->get_temp_recovery_object(recovery_info.version,
```

Comment:
I'd rather bypass the whole thing.  I'm not really worried making sure we capture the sparse behavior (you'll notice that they current implementation doesn't anyway).  In the scenario you are describing, 0M~1M will show up as unchanged, so we will simply recover 1M~4M (much of which will be zeros).  Of course, you need to make sure do_osd_ops appropriately adjusts the clean regions during truncate (I assumed it did, but can't remember).